### DISCUSSION TITLE
How to get ancestorsAnd self for multiple records at once ?

### DISCUSSION THREAD

**Discussion by cmendez2121** [staudenmeir/laravel-adjacency-list#251] [Q&A] [answered]
HI i have an issue selecting nodes with Ancestors and make it as a tree estructure 

4 and 8 are leaf  nodes / records 

$cuentas = CatCuentas::find([4, 8])->ancestorsAndSelf()->depthFirst()->get();

$cuentas = CatCuentas::whereIn('id',[4,8])->descendantsAndSelf()->get();

and get Method Staudenmeir\LaravelAdjacencyList\Eloquent\Collection::ancestorsAndSelf does not exist.

if i do $cuentas = CatCuentas::find(4)->ancestorsAndSelf()->depthFirst()->get();

Works fine for a singe record.  Id 4 is leaf. 
but how to make it work for multiple records and get Tree ordered by depth  (1,2,3,...) 
if i do this 

`
        $constraint = function ($query) {
            $query->whereIn('id', [4, 8]);
        };
        $cuentas = CatCuentas::with('ancestors')->treeOf($constraint)->get();
`
i get the node 4  

`
{
    "id": 4,
    "parent_id": 3,
    "level": 3,
    "name": "Test node  1-3-4",
   "ancestors": [
      {
        "id": 3,
        "parent_id": 1,
        "level": 2,
        "name" :  "Test node  1-3",
      },
      {
        "id": 1,
        "parent_id": null,
        "level": 1,
        "name" :  "Test node  1"
      }
    ]
}
`

but its not the desired result , any pointers ?

**Comment by staudenmeir** [comment]
Hi @cmendez2121,

> but how to make it work for multiple records and get Tree ordered by depth (1,2,3,...)

Do you want to get the part of tree that only contains the paths (from root to leaf) of these leaf records? Do these leaf records always lead to the same root node?

**Reply by cmendez2121** [reply]
HI, i want to get all the records (part of the tree) and make the tree estructure  so at the end of the query make 

 $cuentas->toTree();

-- Root node  (id 1)  
       + children : --- Child node - id(3) 
               + children  ----- Childnode.  (4)  [ in the  origin of search]

--- root node  id 6 
       +. children : ---  child node - id (8)  [in the origin of search]

**Comment by staudenmeir** [comment, marked as answer]
I don't see an elegant solution for this.

You need to get the ancestors separately and normalize their depths so that all root nodes have the depth `0`. Then you can merge them into a single tree:

```php
$a = CatCuentas::find(4)->ancestorsAndSelf;
$a = $a->each(fn($model) => $model->depth -= $a->min('depth'));

$b = CatCuentas::find(8)->ancestorsAndSelf;
$b = $b->each(fn($model) => $model->depth -= $b->min('depth'));

$tree = $a->merge($b)->toTree();
```

**Reply by cmendez2121** [reply]
HI @staudenmeir , i was making more tests and all were leading me to the one you describe it, 
would be nice in the future to have a feature for multiple models /  records to achieve this
thanks for the tip
